tools/base/aio.py

```diff
@@ -83,3 +105,386 @@ async def run_system_command():
+                    yield coro
+            else:
+                for coro in self._coros:  # type:ignore
+                    yield coro
+        except BaseException as e:
+            # Catch all errors iterating (other errors are caught elsewhere)
+            # If iterating raises, wrap the error and send it to `submit` and
+            # and `output` to close the queues.
+            yield AsyncParallelIteratorError(e)
+
+    async def on_task_complete(self, result: Any, decrement: Optional[bool] = True) -> None:
+        """Output the result, release the sem lock, decrement the running
+        count, and notify output queue if complete.
+        """
+        if self.closed:
+            # Results can come back after the queue has closed as they are
+            # cancelled.
+            # In that case, nothing further to do.
+            return
+
+        # Give result to output
+        await self.out.put(result)
+
+        if not self.nolimit:
+            # Release the sem.lock
+            self.sem.release()
+        if decrement:
+            # Decrement the running_queue if it was incremented
+            self.running_queue.get_nowait()
+        # Exit if nothing left to do
+        await self.exit_on_completion()
+
+    async def output(self) -> AsyncIterator:
+        """Asynchronously yield results as they become available"""
+        while True:
+            # Wait for some output
+            result = await self.out.get()
+            if result is _marker:
+                # All done!
+                await self.close()
+                break
+            elif self.should_error(result):
+                # Raise an error and bail!
+                await self.cancel()
+                raise result
+            yield result
+
+    async def ready(self) -> bool:
+        """Wait for the sem.lock and indicate availability in the submission
+        queue
+        """
+        if self.closed:
+            return False
+        if not self.nolimit:
+            await self.sem.acquire()
+            # We check before and after acquiring the sem.lock to see whether
+            # we are `closed` as these events can be separated in
+            # time/procedure.
+            if self.closed:
+                return False
+        return True
+
+    def remember_task(self, task: asyncio.Task) -> None:
+        """Remember a scheduled asyncio task, in case it needs to be
+        cancelled
+        """
+        self.running_tasks.append(task)
+        task.add_done_callback(self.forget_task)
+
+    def should_error(self, result: Any) -> bool:
+        """Check a result type and whether it should raise an error"""
+        return (
+            isinstance(result, AsyncParallelIteratorError)
+            or (isinstance(result, AsyncParallelError) and not self.yield_exceptions))
+
+    async def submit(self) -> None:
+        """Process the iterator of coroutines as a submission queue"""
+        await self.submission_lock.acquire()
```

Comment:
For my education, can you manage asyncio locks with context managers?

Reply:
yep - good q - you can - much like temporary directories etc

the reason for not doing so here, is that in a class its easier to hold the reference as an instance var and cleanup after, than having to know which bits of code are in/outside the context, and having to pass around any context vars

Reply:
as a follow up to this i have a class that wants to be iterated but also needs some cleanup/exit code 

for this we can combine context and generators as follows

```python


class Foo:

    def __enter__(self):
        return self

    def __exit__(self, *args):
        # do cleanup etc
        pass

    async def __aiter__(self):
        with self:
            for x in range(0, 5):
                yield x
```

which can be used like:

```
async for result in Foo():
    print(result)
```